### ISSUE TITLE
MCP23017: during software reset of an ESP the outputs are pulled LOW for a short time

### ISSUE THREAD

**Issue by gsieben** [esphome/issues#3444]
### The problem

The MCP23017 has the characteristic to pull pins defined as outputs low for a short time period when resetting/flashing/rebooting the ESP32/ESP8266. This is a problem when using low-active relays. This is because the relays are briefly switched to active on each boot. The rapid clicking of all relays is easily heard when using mechanical relays.

An easy workaround is to connect the MCP23017 to the reset pin of the ESP. Then the MCP23017 is reset and the outputs are not pulled to LOW. But this only works the reset button is pressed. Not if a software reset is performed

Possibly this is a problem of the MCP23017 library. 

### Which version of ESPHome has the issue?

2022.6.2

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.7.4

### What platform are you using?

ESP32

### Board

ESP32-DevKitC

### Component causing the issue

MCP23017

### Example YAML snippet

```yaml
Not relevant
```

### Anything in the logs that might be useful for us?

```txt
Not relevant
```

**Comment by pmannk** [comment]
You might want to take a look at this (closed) issue. early_pin_init might help you, though there are also hardware characteristics of ESP devices which you need to take into account.
https://github.com/esphome/issues/issues/3280

**Comment by gsieben** [comment]
I will have a look at it. Thank you very much. Soon I will have vacation. Have also read something that one can arrange a software reset. But what I read was too unspecific. Currently I have solved it with high-active relay.

**Comment by VarunKumaran** [comment]
> ### The problem
> The MCP23017 has the characteristic to pull pins defined as outputs low for a short time period when resetting/flashing/rebooting the ESP32/ESP8266. This is a problem when using low-active relays. This is because the relays are briefly switched to active on each boot. The rapid clicking of all relays is easily heard when using mechanical relays.
> 
> An easy workaround is to connect the MCP23017 to the reset pin of the ESP. Then the MCP23017 is reset and the outputs are not pulled to LOW. But this only works the reset button is pressed. Not if a software reset is performed
> 
> Possibly this is a problem of the MCP23017 library.
> 
> ### Which version of ESPHome has the issue?
> 2022.6.2
> 
> ### What type of installation are you using?
> Home Assistant Add-on
> 
> ### Which version of Home Assistant has the issue?
> 2022.7.4
> 
> ### What platform are you using?
> ESP32
> 
> ### Board
> ESP32-DevKitC
> 
> ### Component causing the issue
> MCP23017
> 
> ### Example YAML snippet
> ```yaml
> Not relevant
> ```
> 
> ### Anything in the logs that might be useful for us?
> ```
> Not relevant
> ```
> 
> ### Additional information
> _No response_

Hii i am having the same issue here... What do you mean by connecting mcp23017 to the reset pin of the esp ?? Does it solve the problem or do we need to use high trigger relay to solve this problem??

**Comment by VarunKumaran** [comment]
> I will have a look at it. Thank you very much. Soon I will have vacation. Have also read something that one can arrange a software reset. But what I read was too unspecific. Currently I have solved it with high-active relay.

Another solution is to use a simple external optocoupler (pc817) to make the low active relay turn on when there is high signal from mcp23017... Kind of making the low active relay as high active relay

**Comment by gsieben** [comment]
Yes, you can certainly solve this with a circuit. The circuit is trivial. But it's no fun with 8 or 16 outputs and that's just a hardware hack. And it's bad because it's cumbersome. The usual relays on the market are low-active. Many will have or get this problem. I am sure there is a software solution. I will have a look at the code soon. 

The reset pin didn't really work unfortunately. Because when you press the reset head on the ESP board (if the board has one), it pulls the ESP reset pin and the MCP23017 reset pin low. This puts both into reset mode. And the MCP23017 does not switch at restart. But if you do a software reset (or OTA update), the reset pin of the ESP is not pulled low and therefore the MCP23017 is not set to reset mode.

**Comment by ssieb** [comment]
This should be fixed by https://github.com/esphome/esphome/pull/3744
